hydra/errors.py

```diff
@@ -39,3 +39,7 @@ def __init__(
         super(MissingConfigException, self).__init__(message)
         self.missing_cfg_file = missing_cfg_file
         self.options = options
+
+
+class HydraDeprecationError(HydraException):
+    ...
```

Comment:
I wasn't sure whether to inherit from `HydraException` or from `CompactHydraException`.

Inheriting from `HydraException`, we get this traceback:
```text
$ HYDRA_DEPRECATION_WARNINGS_AS_ERRORS=1 python tests/test_apps/deprecation_warning/my_app.py
Error executing job with overrides: []
Traceback (most recent call last):
  File "/home/jbss/hydra.git/tests/test_apps/deprecation_warning/my_app.py", line 10, in my_app
    deprecation_warning("Feature FooBar is deprecated")
  File "/home/jbss/hydra.git/hydra/_internal/deprecation_warning.py", line 11, in deprecation_warning
    raise HydraDeprecationError(message)
hydra.errors.HydraDeprecationError: Feature FooBar is deprecated

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```
Inheriting from `CompactHydraException`, we get this output:
```text
$ HYDRA_DEPRECATION_WARNINGS_AS_ERRORS=1 python tests/test_apps/deprecation_warning/my_app.py
Error executing job with overrides: []
Feature FooBar is deprecated

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```
@omry and @jieru-hu, which is preferable?

Reply:
Can you try to trigger a realistic deprecation warning a user would see? say if they forget to specify config_path.

Reply:
Forgetting to specify config_path like this:
```python
import hydra
@hydra.main()
def my_app(cfg):
    pass
my_app()
```
Gives the same traceback for both `HydraException` and `CompactHydraException`:

```text
Traceback (most recent call last):
  File "/home/jbss/hydra.git/my_app.py", line 2, in <module>
    @hydra.main()
  File "/home/jbss/hydra.git/hydra/main.py", line 29, in main
    deprecation_warning(
  File "/home/jbss/hydra.git/hydra/_internal/deprecation_warning.py", line 11, in deprecation_warning
    raise HydraDeprecationError(message)
hydra.errors.HydraDeprecationError:
config_path is not specified in @hydra.main().
See https://hydra.cc/docs/next/upgrades/1.0_to_1.1/changes_to_hydra_main_config_path for more information.
```
I think we are getting the same traceback in this case because the `deprecation_warning` is triggered before `run_job` gets called; the exception-sanitizing logic in `run_and_report` looks for the word `"run_job"` in the traceback.

Triggering a deprecation warning using `TargetConf`:
```python
from hydra import main, types
@main(config_path=None)
def my_app(cfg):
    _ = types.TargetConf()
my_app()
```
Here is the output if we inherit from `HydraException`:
<details><summary>(long traceback)</summary>

```text
$ HYDRA_DEPRECATION_WARNINGS_AS_ERRORS=1 python my_app.py
Error executing job with overrides: []
Traceback (most recent call last):
  File "/home/jbss/hydra.git/hydra/_internal/utils.py", line 211, in run_and_report
    return func()
  File "/home/jbss/hydra.git/hydra/_internal/utils.py", line 368, in <lambda>
    lambda: hydra.run(
  File "/home/jbss/hydra.git/hydra/_internal/hydra.py", line 110, in run
    _ = ret.return_value
  File "/home/jbss/hydra.git/hydra/core/utils.py", line 233, in return_value
    raise self._return_value
  File "/home/jbss/hydra.git/hydra/core/utils.py", line 160, in run_job
    ret.return_value = task_function(task_cfg)
  File "/home/jbss/hydra.git/my_app.py", line 4, in my_app
    _ = types.TargetConf()
  File "<string>", line 4, in __init__
  File "/home/jbss/hydra.git/hydra/types.py", line 38, in __post_init__
    deprecation_warning(message=msg)
  File "/home/jbss/hydra.git/hydra/_internal/deprecation_warning.py", line 11, in deprecation_warning
    raise HydraDeprecationError(message)
hydra.errors.HydraDeprecationError:
TargetConf is deprecated since Hydra 1.1 and will be removed in Hydra 1.2.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jbss/hydra.git/my_app.py", line 5, in <module>
    my_app()
  File "/home/jbss/hydra.git/hydra/main.py", line 48, in decorated_main
    _run_hydra(
  File "/home/jbss/hydra.git/hydra/_internal/utils.py", line 367, in _run_hydra
    run_and_report(
  File "/home/jbss/hydra.git/hydra/_internal/utils.py", line 251, in run_and_report
    assert mdl is not None
AssertionError
```
</details>
And here is the output if we inherit from `CompactHydraException`:

```text
$ HYDRA_DEPRECATION_WARNINGS_AS_ERRORS=1 python my_app.py
Error executing job with overrides: []

TargetConf is deprecated since Hydra 1.1 and will be removed in Hydra 1.2.

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```

Reply:
In my opinion the `CompactHydraException` is much cleaner, but there is a problem:
someone might be confused if the output just says "Foobar is deprecated", since normally deprecations do not cause a program to fail immediately.

What do you think about adding the word `"HydraDeprecationError"` to the output? In the `deprecation_warning` function, we could replace
```python
raise HydraDeprecationError(message)
```
with

```python
raise HydraDeprecationError("HydraDeprecationError: " + message)
```
to produce this output (inheriting from `CompactHydraException`):

```text
$ HYDRA_DEPRECATION_WARNINGS_AS_ERRORS=1 python my_app.py
Error executing job with overrides: []
HydraDeprecationError:
TargetConf is deprecated since Hydra 1.1 and will be removed in Hydra 1.2.

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```

Reply:
Deprecation warnings typically tell you where the problem is. I don't we should keep the more verbose version here.
This is just a debugging tool and not something users will constantly use.

Reply:
Ok.